src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+    /// Upon expiration of the defined <c>period</c>, a full reload is triggered.
+    /// Unless <c>LoadOption.AllowStale</c> is used, each cache hit still incurs a roundtrip to load any subsequently-added events.
+    | FixedTimeSpan of ICache * period : TimeSpan
+    /// Prefix is used to segregate multiple folds per stream when they are stored in the cache.
+    /// Semantics are identical to <c>SlidingWindow</c>.
+    | SlidingWindowPrefixed of ICache * window : TimeSpan * prefix : string
+
+type MessageDbCategory<'event, 'state, 'context>(resolveInner, empty) =
+    inherit Equinox.Category<'event, 'state, 'context>(resolveInner, empty)
+    new (   context : MessageDbContext, codec : IEventCodec<_, _, 'context>, fold, initial,
+            [<O; D(null)>]?caching,
+            [<O; D(null)>]?access) =
+        let inner = Category<'event, 'state, 'context>(context, codec, ?access = access)
+        let readCacheOption =
+            match caching with
+            | None -> None
+            | Some (CachingStrategy.SlidingWindow (cache, _))
+            | Some (CachingStrategy.FixedTimeSpan (cache, _)) -> Some (cache, null)
+            | Some (CachingStrategy.SlidingWindowPrefixed (cache, _, prefix)) -> Some (cache, prefix)
+        let folder = Folder<'event, 'state, 'context>(inner, fold, initial, ?readCache = readCacheOption)
+        let category : ICategory<_, _, 'context> =
+            match caching with
+            | None -> folder :> _
+            | Some (CachingStrategy.SlidingWindow (cache, window)) ->
+                Caching.applyCacheUpdatesWithSlidingExpiration cache null window folder Token.supersedes
+            | Some (CachingStrategy.FixedTimeSpan (cache, period)) ->
+                Caching.applyCacheUpdatesWithFixedTimeSpan cache null period folder Token.supersedes
+            | Some (CachingStrategy.SlidingWindowPrefixed (cache, window, prefix)) ->
+                Caching.applyCacheUpdatesWithSlidingExpiration cache prefix window folder Token.supersedes
+        let resolveInner streamIds = struct (category, StreamName.Internal.ofCategoryAndStreamId streamIds, ValueNone)
+        let empty = struct (context.TokenEmpty, initial)
+        MessageDbCategory(resolveInner, empty)
+
+
+type MessageDbConnector(
+    connectionString : string, ?readOnlyConnectionString : string,
+    [<O; D(null)>]?readRetryPolicy, [<O; D(null)>]?writeRetryPolicy) =
+        member _.Connect() =
+            let connectToDb connectionString ct = task {
+                let conn = new NpgsqlConnection(connectionString)
+                do! conn.OpenAsync(ct)
+                return conn }
+            let writeConnection = connectToDb connectionString
+            let readConnection =
+                match readOnlyConnectionString with
+                | Some connectionString -> connectToDb connectionString
+                | None -> writeConnection
+            MessageDbClient(writeConnection), MessageDbClient(readConnection)
```

Comment:
do you need 2 separate clients instances? i.e. could wrapping not happen on L425?

Reply:
I must admit I'm not entirely sure what you're getting at.

I've simplified this a bit but I think the answer is yes, we need two instances.

Reply:
if the read and write conn string are the same, then the read client should be the same instance as the write, no need to construct two (but solving the deeper issue of duplication of duties between client and connection will probably remove this concern entirely)